### PR TITLE
uuid5

### PR DESCRIPTION
**Pull request by matbesancon**
Create UUIDs version 5 using the SHA1 hashing of another UUID (namespace) and string (name).
Special thanks to @ScottPJones for helping out with the last bits :trollface:

### CODE REVIEW

File: stdlib/UUIDs/src/UUIDs.jl [hunk 1]
```diff
@@ -81,4 +83,34 @@ function uuid4(rng::AbstractRNG=Random.GLOBAL_RNG)
...
+```
+"""
+function uuid5(ns::UUID, name::String)
+    hash_result = SHA.sha1(string(ns,name))
+    # set version number to 5
+    hash_result[7] = (hash_result[7] & 0x0F) | (0x50)
+    hash_result[9] = (hash_result[9] & 0x3F) | (0x80)
+    v = zero(UInt128)
+    #use only the first 16 bytes of the SHA1 hash
+    for idx in Base.OneTo(16)
```
Comment (ararslan) [inline stdlib/UUIDs/src/UUIDs.jl]: Why not just `1:16`?

**Comment by iamed2** [comment]
I suggest including the [provided namespaces](https://tools.ietf.org/html/rfc4122.html#appendix-C) from the RFC as constants, and testing with UUID5s generated by, e.g., Python (which also provides the namespaces from the RFC as constants).

File: stdlib/UUIDs/src/UUIDs.jl (hunk 1 above)
Comment (matbesancon) [reply stdlib/UUIDs/src/UUIDs.jl]: OneTo is generally, in some applications lots of UUIDs can get generated.
I'll build a benchmark if the argument is needed

**Comment by matbesancon** [comment]
I tried comparing the result of the Python lib with the current implementation and of course it's different, I'll look into it

**Comment by matbesancon** [comment]
The results produced on the standard namespaces are the same as the rust ones:
```rust
extern crate uuid;
use uuid::Uuid;

fn main() {
    let refs = vec![&uuid::NAMESPACE_DNS, &uuid::NAMESPACE_OID, &uuid::NAMESPACE_URL, &uuid::NAMESPACE_X500];
    let pairs = refs
        .iter()
        .map(|u|
            (u, Uuid::new_v5(u, "julia"))
        );
    for (u1,u2) in pairs {
        println!("{0} - {1}", u1, u2);
    }
}
```

**Comment by iamed2** [comment]
Excellent :)

**Comment by matbesancon** [comment]
ok "we" have one tiny problem, `uuidgen` just got v5 on Ubuntu18.04 (not sure when it was integrated) and yields a slightly different result at the 17th byte for **some** results, not all. The results in the lib are still consistent with Python and Rust, just the OS uuidgen yields something different.
```bash
$ uuidgen --sha1
$ uuidgen --sha1 --namespace @dns --name julia
00ca23ad-40ef-500c-8910-157de3950d07
$ uuidgen --sha1 --namespace @oid --name julia
b7bf72b0-fb4e-538b-952a-3be296f07f6d
$ uuidgen --sha1 --namespace @url --name julia
997cd5be-4705-5439-9fe6-d77b18d612e5
$ uuidgen --sha1 --namespace @x500 --name julia
993c6684-82e7-5cdb-9d46-9bff0362e6a9
```

Python (same as Julia and Rust)
```python
In [1]: import uuid
In [2]: uuid.uuid5(uuid.NAMESPACE_DNS,"julia")
Out[2]: UUID('00ca23ad-40ef-500c-a910-157de3950d07')
In [3]: uuid.uuid5(uuid.NAMESPACE_OID,"julia")
Out[3]: UUID('b7bf72b0-fb4e-538b-952a-3be296f07f6d')
In [4]: uuid.uuid5(uuid.NAMESPACE_URL,"julia")
Out[4]: UUID('997cd5be-4705-5439-9fe6-d77b18d612e5')
In [5]: uuid.uuid5(uuid.NAMESPACE_X500,"julia")
Out[5]: UUID('993c6684-82e7-5cdb-bd46-9bff0362e6a9')
```

**Comment by StefanKarpinski** [comment]
Sigh. It seems like matching Python and Rust is the way to go.

**Comment by iamed2** [comment]
I believe this indicates the [variant](https://en.wikipedia.org/wiki/Universally_unique_identifier#Variants), and I think you can ignore that digit.

**Comment by iamed2** [comment]
(they shouldn't be different though, as both Python and Linux' uuid-gen claim to follow RFC-4122)

**Comment by matbesancon** [comment]
Maybe the variant is different though (different variant are specified by
the RFC) and it seems to match the byte change location. The weird thing
is, why is it just sometimes different?

**Comment by iamed2** [comment]
Ohhh, I see:

From the RFC:
> The following table lists the contents of the variant field, where the letter "x" indicates a "don't-care" value.
> 
>     1     0     x    The variant specified in this document.

So `b` and `9` (`1011` and `1001`) should be considered equivalent, same with `a` and `8` (`1010` and `1000`).

**Comment by matbesancon** [comment]
thanks for clarifying this, I was trying to find where to file a bug for uuidgen

**Comment by iamed2** [comment]
No problem, stuff like this eats at me until I figure it out 😄

**Comment by matbesancon** [comment]
should this get the 1.0 flag, since this is extending the exposed API of UUID?

**Comment by StefanKarpinski** [comment]
Yes, that's correct. It's a feature so it goes in a minor release.

**Pull request merged**